compose: Handle *general chat* in topic box.

prakhar1144 opened the pull request:
#33338

* Added last two commits.
* Updated the third commit: Placeholder should live-update when `mandatory_topics` setting is changed.
* Minor fix in the fourth commit: `is_empty_stream_topic` -> `is_empty_string_topic`
* Pulled a commit from #33311 
* Simplified 3rd last commit. (Discussed with Karl in DM before making changes)

<img width="462" alt="Screenshot 2025-02-13 at 8 24 09 PM" src="https://github.com/user-attachments/assets/c5727bff-96c7-40e4-a9ce-c0cb885ac651" />


---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
If I click on the channel picker an then the main compose area, the topic typeahead blinks for a moment.

alya commented:
I don't know if it's related to this PR, but sometimes the typeahead goes downwards even thought there's not enough space for it.

![Screenshot 2025-02-13 at 09 03 42@2x](https://github.com/user-attachments/assets/495893f8-6e1a-4283-85c3-f646c8df871e)

prakhar1144 commented:
Fixed the following in the recent push:
* https://github.com/zulip/zulip/pull/33465#issuecomment-2657216068
* [CZO bug report](https://chat.zulip.org/#narrow/channel/101-design/topic/general.20chat.20.2F.20default.20topic/near/2090503)
* [CZO bug report](https://chat.zulip.org/#narrow/channel/9-issues/topic/leftover.20topic.20typeahead/near/2091419) -- I wasn't able to replicate this but I'm more confident that this bug would be gone as I reworked with a new approach.

alya commented:
@timabbott Let me know if you need me to re-test.

alya commented:
@timabbott  and I noticed in manual testing that we incorrectly show the "messages are faded" banner here:

![Screenshot 2025-02-18 at 10 12 31](https://github.com/user-attachments/assets/dfaa85aa-0194-40b0-8e78-a8907c0a0a5c)

The banner should only appear once you've existed the topic box, like it does with mandatory topics.

alya commented:
In general, I'm really disliking showing banners behind the typeahead. I'm not sure how easy that is to fix for the general case.

timabbott commented on web/src/composebox_typeahead.ts:
```diff
@@ -954,9 +955,17 @@ export function get_candidates(
+                const empty_string_topic_idx = topic_list.indexOf("");
+                if (empty_string_topic_idx !== -1) {
+                    topic_list[empty_string_topic_idx] = empty_string_topic_display_name;
+                }
```
I feel like this could be cleaned up by just having `get_topic_matcher` do `util.get_final_topic_display_name`, but maybe I'm missing something.

```
`function get_topic_matcher(query: string): (topic: string) => boolean {                                 
    query = typeahead.clean_query_lowercase(query);                                                     
                                                                                                        
    return function (topic: string): boolean {                                                          
        return typeahead.query_matches_string_in_order(query, topic, " ");                              
    };                                                                                                  
}                                                                                                       
```

prakhar1144 replied:
Yeah, that's a cleaner approach.

Made the changes.

timabbott commented on web/src/composebox_typeahead.ts:
```diff
@@ -1231,10 +1241,14 @@ export function content_typeahead_selected(
-                topic_link_util.get_stream_topic_link_syntax(stream_name, item.topic) +
-                " ";
+            const stream_topic_link_syntax =
+                item.topic_display_name === util.get_final_topic_display_name("")
```
In think ideally, we'd have the original `topic` on the object, and be comparing `topic === ""`, but I'm not sure it's very important.

timabbott replied:
It probably doesn't matter, this can be as-is.

timabbott commented on web/src/compose_recipient.ts:
```diff
@@ -359,10 +359,14 @@ export let update_compose_area_placeholder_text = (): void => {
             message_type,
             stream_id,
-            topic: compose_state.topic(),
+            topic,
```
Hmm, would it make sense for `compose_ui.compute_placeholder_text` to be responsible for doing this translation? I think that'd be a better abstraction, but it should have no functional impact.

prakhar1144 replied:
Yeah.

Actually, I ended up dropping Karl's commit as it was a bit buggy -- start the compose box in CZO using "start new conversation" button or something, the placeholder shows "test general chat".

Added a commit which fixes this bug and doesn't require passing `empty_topic_display` parameter around.

timabbott commented on web/src/compose_fade_helper.ts:
```diff
@@ -31,11 +32,8 @@ export function want_normal_display(): boolean {
-        // (which is most realms as of this writing).
-        if (focused_recipient.topic === "") {
+        // If the topic is empty, and the realm requires topics, then we want a normal display.
+        if (focused_recipient.topic === "" && realm.realm_mandatory_topics) {
```
I think this is the fix for the banner appearing underneath.

``` diff
diff --git a/web/src/compose_fade_helper.ts b/web/src/compose_fade_helper.ts
index 414665ef4f..e903fc6661 100644
--- a/web/src/compose_fade_helper.ts
+++ b/web/src/compose_fade_helper.ts
@@ -33,7 +33,10 @@ export function want_normal_display(): boolean {
         }
 
         // If the topic is empty, and the realm requires topics, then we want a normal display.
-        if (focused_recipient.topic === "" && realm.realm_mandatory_topics) {
+        if (
+            focused_recipient.topic === "" &&
+            (realm.realm_mandatory_topics || $("input#stream_message_recipient_topic").is(":focus"))
+        ) {
             return true;
         }
     }

```

prakhar1144 replied:
Yeah, this fix is correct. But there's another problem:

We want to `update_fade` and show banner when user clicks on compose text-area. 

We have:
```
    $("textarea#compose-textarea").on("focus", () => {
        compose_recipient.update_compose_area_placeholder_text();
        if (narrow_state.narrowed_by_reply()) {
            compose_notifications.maybe_show_one_time_non_interleaved_view_messages_fading_banner();
        } else {
            compose_notifications.maybe_show_one_time_interleaved_view_messages_fading_banner();
        }
    });
 ```
 
We'll add `compose_recipient.update_fade()` here and then `...fading_banner` function will be called.

The problem is, the `...fading_banner` checks if the fading is done but we add throttle to fading. So, the banner won't be visible.

timabbott replied:
I think we could just call `do_update_all` instead here, to skip the throttling? The comment is clear it's intended for keyup events.

prakhar1144 replied:
Ah, yes, `do_update_all` will do the job. Done -- #33588

timabbott commented on web/src/composebox_typeahead.ts:
```diff
@@ -1411,6 +1411,14 @@ export function initialize_compose_typeahead($element: JQuery<HTMLTextAreaElemen
+    if (!realm.realm_mandatory_topics && $("input#stream_message_recipient_topic").val() === "") {
+        composebox_topic_typeahead.lookup(false);
+    }
+}
```
I'm not sure where in the code the bug is, but the `C` keyboard shortcut isn't working properly in the development environment for me; it seems to start in general chat, rather than starting a new topic.

![image](https://github.com/user-attachments/assets/bec4b871-94e0-467f-9927-42ef32cdac0b)

prakhar1144 replied:
```diff
-        if (opts.trigger === "clear topic button") {
+        if (opts.trigger === "clear topic button" || opts.trigger === "compose_hotkey") {
            return "input#stream_message_recipient_topic";
        }
```

Had to add this to fix.

timabbott replied:
Yeah, that sounds right.

timabbott commented:
I think that's a complete set of feedback. I'll update the chat.zulip.org deployment.

prakhar1144 commented:
Addressed the review comments.

Posted a question https://github.com/zulip/zulip/pull/33465#discussion_r1963486718

timabbott commented:
I accidentally passed the wrong PR ID when merging https://github.com/zulip/zulip/pull/33577 via `push-to-pull-request --merge`, so we need to make a new PR for this.
